src/fsharp/service/service.fsi

```diff
@@ -316,7 +317,7 @@ type public FSharpChecker =
     /// TypeCheck and compile provided AST
     /// </summary>
     /// <param name="userOpName">An optional string used for tracing compiler operations associated with this request.</param>
-    member CompileToDynamicAssembly: ast:ParsedInput list * assemblyName:string * dependencies:string list * execute:(TextWriter * TextWriter) option * ?debug:bool * ?noframework:bool  * ?userOpName: string -> Async<FSharpErrorInfo [] * int * System.Reflection.Assembly option>
+    member CompileToDynamicAssembly: ast:ParsedInput list * assemblyName:string * dependencies:string list * execute:(TextWriter * TextWriter) option * ?debug:bool * ?primaryAssembly:PrimaryAssembly * ?noframework:bool * ?userOpName: string -> Async<FSharpErrorInfo [] * int * System.Reflection.Assembly option>
```

Comment:
To Consider:
* Should we expose `PrimaryAssembly` or use a different parameter? `PrimaryAssembly` is awkwardly located in the `FSharp.Compiler.AbstractIL` which has to be opened as well.
* What should the default value be? Currently it just uses the `TcConfigB` default which is mscorlib. However I'm wondering if using a default based on the current runtime might be more sensible.

Reply:
Yeah the default should be determined by TFM.

Generally I wouldn't like to expose this info since it feels very much like an implementation detail. I removed some IVTs that snagged items from absil and although this is a better approach it'd be nice if it weren't necessary.

Also idk what a PrimaryAssembly is

Reply:
> Also idk what a PrimaryAssembly is

It corresponds to the `--targetprofile` argument accepted by the compiler. It has three possible values: `mscorlib`, `netstandard` or `netcore`. The TcConfig default is `mscorlib` regardless of tfm or current runtime.

Reply:
So to answer your questions, I think it should be exposed, since the cli does the same. The question however is whether we should use a special type for the FsharpChecker API and map to PrimaryAssembly under the hood.

Reply:
Gotcha. This seems fine then, since it's a CLI option.

Reply:
@cartermp this primaryassembly option is actually the thing o fixed that fixed FSAC FSI typechecking. It wasn't obvious to me because we use ProjectOptions, where this is hidden behind the --targetprofile flag. So I'm 💯 in support of making this API a nice strong DU.

Reply:
I've been thinking about this.

It turns out we won't be able to remove the `targetprofile` command line option from Fsc as existing calls to the command line compiler rely on it.

However, exposing the notion of a `targetprofile` as part of an API to the compiler is something I would like to avoid; the core compiler itself really should not care about the profile (even though it does today, but Fsc (CLI compiler) is really the only one that should care). If the reasoning is, the "CLI" does it, then everything that the "CLI" allows should be exposed in some way. Perhaps the APIs from Compile/CompileDynamicAssembly that takes ASTs should also take in CLI args and we simply just ignore any source files listed or see if the source files listed match the ASTs' file name.

Reply:
> Perhaps the APIs from Compile/CompileDynamicAssembly that takes ASTs should also take in CLI args

I'm assuming you're approaching this from the point of view of breaking changes? I was thinking the same, however for better or for worse that particular subset of the api is accepting individual arguments for things like `--no-framework`, so requiring that suddenly become part of an argv is also a breaking change.

> However, exposing the notion of a targetprofile as part of an API to the compiler is something I would like to avoid

I see your point but for all intents and purposes it is exposed. The goal of this PR is to fix something that is currently not working and missing acceptance tests for it. My goal here is to be unblocked for QuotationCompiler. Should this change in the future, changing the signature should not be an issue, provided the tests still pass.